### PR TITLE
Resource estimator 2.0

### LINKED ISSUE #37338: Resource Estimator: Update existing calculator

### PR DESCRIPTION
**Pull request by abeatrix**
# Summary

Close https://github.com/sourcegraph/sourcegraph/issues/37338

This PR is for rewriting the resource estimator as one of the tiger team projects.

It aims to meet the following requirements for the project:
- [x] Ensure all services are included 
- [x] Use the most up-to-date data retrieved from [this spreadsheet](https://docs.google.com/spreadsheets/d/1N7X_OXDwKk0QSR2Ghbj7ZhjVrQXcMNj-yC8mF1amBi4/edit#gid=609023588)
- [x] Provide recommendations/guidance for database selection, configuration, and disk guidance.

- It also asks for more information to provide a more accurate estimation
- Remove redundant information
- Add explanation for commonly confused terms, ex engagement rate, default deployment size
- Add an overview for each service
- Add estimation for storage size
- Add estimation for ephemeral storage size

## Known limitation

- the estimator is based on the very limited data we have
- some of the data include deployment configurations that are overprovisioned
- the estimator does not work for deployments that have more than:
  - 10k users
  - 50k repo
  - 20 repos larger than 1GB
  - 5TB for total repo size

### Calculator

image

### If an input exceeds a limit

image

### Service Overview under the calculator

image

### Additional Information

image

image

### Future Plan

I'd like to rewrite the estimator with TypeScript as the current format (Markdown) has a lot of limitations

### CODE REVIEW

**Review by emidoots — APPROVED** [review]
generally LGTM (even if not ready yet)

File: internal/scaling/scaling.go
Comment (caugustus-sourcegraph) [inline internal/scaling/scaling.go:424]: For gitserver / indexed-search: is the size shown per replica, or in total? I'm guessing it's total and the customer will want to divide by the number of replicas when setting their PVC sizes in Kubernetes. Any concerns with that approach?

File: internal/scaling/scaling.go
Comment (caugustus-sourcegraph) [inline internal/scaling/scaling.go]:
```suggestion
	fmt.Fprintf(&buf, "| searcher| %v | The size of all indexed repos. |\n", fmt.Sprint(float64(e.TotalRepoSize*30/100), "GB"))
```

File: internal/scaling/scaling.go
Comment (caugustus-sourcegraph) [inline internal/scaling/scaling.go]:
```suggestion
	fmt.Fprintf(&buf, "| Service | Description |\n")
```

**Review by caugustus-sourcegraph — APPROVED** [review]
:tada:

File: internal/scaling/scaling.go
Comment (abeatrix) [reply internal/scaling/scaling.go:424]: It did not cross my mind so thanks for pointing this out! Would it help if we add the following footnote:
`This value represents the total disk space required for the associated service. Divide this by the number of replicas to get the value required for each replica.`
image

Comment (caugustus-sourcegraph) [reply internal/scaling/scaling.go:424]: 👍 to the footnote!

This is a concern specific to Kubernetes, so we could change the wording to be even more specific. Maybe something like "For Kubernetes deployments, set the PVC storage size equal to this value divided by the number of replicas."?

If that seems confusing, no need to update - your existing wording is already a good option.

**Comment by keegancsmith** [comment]
🎉 shout if you want extra eyes on the PR :)

**Review by kevinwojo — APPROVED** [review]
This looks much better than I anticipated.

I don't want to block this incremental improvement over an existing gap in the calculator's current implementation.

Customers want to know the CPU and Memory requirements for all services, not just the ones that do change. For instance codeintel-db, minio, exporters. It is redundant since they don't change, but from a planning perspective, they want to know exactly how many compute resources are needed.

Maybe we append static rows to the table for things that don't change?

Again, this is an existing gap in the implementation so I don't believe we should block on this. We can iterate pretty easily on this small gap.

**Comment by abeatrix** [comment]
> 🎉 shout if you want extra eyes on the PR :)

Thanks @keegancsmith , if you have time that'd be awesome. just be warned that my Go is 🤪

(we'd like to push this out asap so I've merged the changes, but if you have any questions or feedbacks for the changes made in this PR please let me know!)

**Pull request merged**
